### DISCUSSION TITLE
Pulumi Linear Mode

### DISCUSSION THREAD

**Discussion by hellt** [pulumi/pulumi#15172] [Ideas]
Hey all,
I have been using pulumi for a couple of weeks and while this sounds like nothing at all, I think this first experience is important to capture for PMs as well as serves as a prelude to this Idea.

Before jumping to the [idea](#idea) itself, I wanted to share my main beefs with the devex of Pulumi.

I've tried to use Go and Py pulumi flavors in a relatively simple use case:

1. **UC1**: Create cloud resources
2. **UC2**: Use resources output values as an input for file generation (yaml manifest)
3. **UC3**: execute another tool that works on that generated yaml to perform config management
4. **UC4**: as a result of 3 make pulumi adopt (fetch/get/lookup) created resources

A seasoned pulumi user might notice a few gotchas in that basic task for an IaC tool.

# P1 Dealing with  outputs

I've spent a couple of days looking at pulumi stack, and one of the most common challenges users face is dealing with outputs outside of their comfort zone.

The problem N1 (P1) is rooted in the async nature of outputs and quite simplistic documentation about it.

UC2 wants to use outputs of several resources (their native values (str, int, etc)) and this becomes a convoluted experience of using `All.Apply` and then dealing with type conversion (for some langs) to get the output value, and then nesting it all in the anonymous func to deal with those values without being able to return the values to the `main` coroutine.

No need to explain this devex is not great, slack conversations confirm that.

# P2 Inter deps between resources and non-resources routines

UC3 wants to use a generated YAML files as an input to the process `foo -f file.yml`.
How do I make pulumi `local` command wait till the yaml is generated. Remember, yaml depends on outputs from resource creation.

It seems the solution is to make the yaml generation func to return a `pulumi.Output`, and then pass this output as an input to a local command using `triggers`.

This. is. not. obvious.

# P3 Waiting for unmanaged resources

This is a small but annoying challenge. If the cloud providers creates a resource based on the resources you created (think of an external LB auto-creation when a service is created in k8s) a nice devex (for me) would be having a function like `lb = pulumi.waitForResource(resourceObj, timeout)`

# Idea

Sorry for the long intro, here is the idea itself. I found that the promise of using a programming lang to manage your infra that pulumi has a face value is great, but it sometimes stands in the way of a developer.

What if we had a **pulumi linear mode** where all resource inter-dependencies are left out for a developer to sort out.

## Benefits

1. **Clear and predictable execution path.** Whatever you have in your `main` executes as if you work with a program written in that language. No magic concurrency, no associated problems with that.
2. **Native types are returned.** Because you don't live in the promise land, resource properties are strings, ints, slices of bytes that you know how to deal with as a developer.

What about concurrency? This is a developer' tasks. They use goroutines, async/await, they have full control over what they want to create concurrently and how.  
Pulumu can probably help with concurrency abstractions that would simplify concurrent resource creation.

This idea might sound crazy, and I might miss some crucial stuff, but I would take this approach quite often with pulumi and it would make my simple IaC programs much cleaner.

# Disclaimer
Hey, and I love pulumi, all of the problems I listed above should be taken as a feedback from a fan who wants to make pulumi better and not an attack.

**Comment by Frassle** [comment]
> This idea might sound crazy

Outputs aren't just for async, they're also the fundamental construct that allows 'preview' to work.
If you were happy with only having `up --skip-preview` then it's technically possible to generate SDKs without any outputs, but this is such a massive limitation we've not entertained the idea.

**Reply by hellt** [reply]
Understood. Then I hope with time, more examples, abstractions, and improvements could come to Outputs workflows, as Scott mentioned [here](https://pulumi-community.slack.com/archives/CCWP5TJ5U/p1705512239663779?thread_ts=1704895154.944869&cid=CCWP5TJ5U), to simplify dealing with them in a more natural (for a chosen language) way.

A couple of people (me included) were thinking that maybe channels are a possible improvement to have the actual value outside of the closure. But again, maybe this breaks preview somehow...

I started writing my IaC in Go, but a dozen resources later and a couple of `All.ApplyT` got me into the mood that I am battling the language instead of enjoying it.
I switched to Py. It got better, but `all.apply` were still there (understandably why).

**Reply by Frassle** [reply]
Go is particularly bad ergonomics because when we built the first SDK it didn't have generics. Outputs are by their nature fully generic types, you get ints, and strings, and lists, and so on. Trying to do this in Go pre-generics meant a lot of repeated boilerplate and explicit casts because the Go type system couldn't express it at all well. We're looking at re-doing outputs in Go with generics but doing this in a way that also maintains compatibility with the existing system has been harder than expected.

Channels don't really help here because as above Outputs are for more than just dealing with async operations but also with preview operations where you might not get a result, e.g. what's the Bucket.Id of a bucket that isn't created yet?
